### PR TITLE
Enhance NumberInput

### PR DESCRIPTION
**Pull request by lbwexler**
This fixes a number of issues with NumberInput. :-)

- Precision now enforced when first focused on overprecise source value.
- Precision now enforced when setting value back to source.
- Debounce provides less thrashing when user typing in numbers
- scaleFactor now clearly documented as being a factor of 10
- Max/min are enforced.  Workaround for related Blueprint errors with this prop handled more reliably 

Also provides some additional template Hooks in HoistInputModel for future use.

Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [X] Caught up with `develop` branch as of last change.
- [X] Added CHANGELOG entry, or determined not required.
- [X] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [X] Updated doc comments / prop-types, or determined not required.
- [X] Reviewed and tested on Mobile, or determined not required.
- [X] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

**Comment by TomTirapani** [comment]
This definitely fixes the issue.

However, I find the new behaviour about discarding out-of-bounds values a bit confusing. There's nothing to indicate *why* it was discarded, and this lack of user feedback, especially when going from a blank input back to a blank input, makes it feel broken.

Couldn't we just remove these props entirely, and say that if your app wants to enforce min/max values to use validation instead?

**Comment by lbwexler** [comment]
Let's discuss -- I do find it useful, especially the ability to have a value of min = 0 or max = 0.  See the usage in LogPanel.

We can recommend validation and forms more strongly in the doc strincs for sure, but I think this is still useful for a raw numberInput.

File: cmp/input/HoistInputModel.js [hunk 1]
```diff
@@ -116,6 +116,7 @@ export class HoistInputModel extends HoistModel {
     @observable.ref internalValue = null;    // Cached internal value
     inputRef = createObservableRef();        // ref to internal <input> element, if any
     domRef = createObservableRef();          // ref to outermost element, or class Component.
+    isDirty = false;
```
Comment (TomTirapani) [inline cmp/input/HoistInputModel.js:119]: I'm not clear on the need for this new flag? Under which circumstances would `doCommitInternal()` be called before `noteValueChange`?

**Comment by TomTirapani** [comment]
Looks great, particularly the precisions enforcing! Eager to get this in 👍

File: cmp/input/HoistInputModel.js (hunk 1 above)
Comment (lbwexler) [reply cmp/input/HoistInputModel.js:119]: when you tab through the form, without making any changes

Comment (lbwexler) [reply cmp/input/HoistInputModel.js:119]: or click in and out of it

**Pull request merged**